src/Rules/Arrays/InvalidKeyInArrayDimFetchRule.php

```diff
@@ -32,12 +37,18 @@ public function processNode(Node $node, Scope $scope): array
 			return [];
 		}
 
-		$varType = $scope->getType($node->var);
-		if (count($varType->getArrays()) === 0) {
+		$dimensionType = $scope->getType($node->dim);
+		$varType = $this->ruleLevelHelper->findTypeToCheck(
+			$scope,
+			$node->var,
+			'',
+			static fn (Type $varType): bool => $varType->isArray()->no() || AllowedArrayKeysTypes::getType()->isSuperTypeOf($dimensionType)->yes(),
+		)->getType();
+
+		if ($varType instanceof ErrorType || $varType->isArray()->no()) {
 			return [];
 		}
 
-		$dimensionType = $scope->getType($node->dim);
```

Comment:
The rest of this method should be:

```php
if ($isSuperType-yes()) {
    return [];
}

return [
	RuleErrorBuilder::message(
		sprintf('%s array key type %s.', $isSuperType->no() ? 'Invalid' : 'Possibly invalid', $dimensionType->describe(VerbosityLevel::typeOnly())),
	)->build(),
];

Reply:
that looks better! adapted. but I had to keep the maybe-early-exit in to avoid the rule reporting maybes if they shouldn't be reported (the newly added levelstest started failing)

Reply:
oh and I forgot to run make phpstan before pushing. need to check what's going on with mixed there which should be ignored still I guess

Reply:
It's a bit weird because RuleLevelHelper should have already taken care of the maybe case on lower levels.

Reply:
Oh, you're right, the rule code needs to be a bit more complicated after all. Not sure what's with the mixed in `make phpstan`, PHPStan is running on level 8...

Reply:
the rule was explicitly ignoring mixed dims before as well when checking for maybe/potential invalid keys. maybe it can't be simplified after all. I was also looking forward to put all checks on top and into the closure

Reply:
the mixed dims that were reported seemed mostly legit, maybe this could even be turned on with bleedingEdge. but it's not related to this PR I think, I let the rule ignore them again. at least it's a bit simpler and makes use of RuleLevelHelper now hmm